src/runtime/common/serialization.cpp

```diff
@@ -0,0 +1,178 @@
+
+#include "hipSYCL/common/serialization.hpp"
+#include "hipSYCL/sycl/access.hpp"
+#include "hipSYCL/sycl/item.hpp"
+#include "hipSYCL/sycl/range.hpp"
+#include "hipSYCL/runtime/backend.hpp"
+#include "hipSYCL/runtime/operations.hpp"
+#include "hipSYCL/runtime/dag.hpp"
+#include "hipSYCL/runtime/device_id.hpp"
+
+#include <iostream>
+#include <map>
+
+
+namespace hipsycl::sycl::access{
+std::ostream& operator<<(std::ostream& out, const mode value){
+    static std::map<mode, std::string> strings;
+    if (strings.size() == 0){
+        strings[mode::read] = "R";
+        strings[mode::write] = "W";
+        strings[mode::atomic] = "atomic";
+        strings[mode::read_write] = "RW";
+        strings[mode::discard_write] = "Discard W";
+        strings[mode::discard_read_write] = "Discard RW";
+    }
+    return out << strings[value];
+}
+
+std::ostream& operator<<(std::ostream& out, const target value){
+    static std::map<target, std::string> strings;
+    if (strings.size() == 0){
+#define INSERT_ELEMENT(p) strings[p] = #p
+        INSERT_ELEMENT(target::image);
+        INSERT_ELEMENT(target::image);
+        INSERT_ELEMENT(target::host_image);
+        INSERT_ELEMENT(target::host_buffer);
+        INSERT_ELEMENT(target::image_array);
+        INSERT_ELEMENT(target::global_buffer);
+        INSERT_ELEMENT(target::constant_buffer);
+#undef INSERT_ELEMENT
+    }
+    return out << strings[value];
+}
+
+std::ostream& operator<<(std::ostream& out, const placeholder value){
+    static std::map<placeholder, std::string> strings;
+    if (strings.size() == 0){
+#define INSERT_ELEMENT(p) strings[p] = #p
+        INSERT_ELEMENT(placeholder::false_t);
+        INSERT_ELEMENT(placeholder::true_t);
+#undef INSERT_ELEMENT
+    }
+    return out << strings[value];
+}
+
+std::ostream& operator<<(std::ostream& out, const fence_space value){
+    static std::map<fence_space, std::string> strings;
+    if (strings.size() == 0){
+#define INSERT_ELEMENT(p) strings[p] = #p
+        INSERT_ELEMENT(fence_space::local_space);
+        INSERT_ELEMENT(fence_space::global_space);
+        INSERT_ELEMENT(fence_space::global_and_local);
+#undef INSERT_ELEMENT
+    }
+    return out << strings[value];
+}
+} // end of hipsycl::sycl::access
+
+
+namespace hipsycl::rt {
+std::ostream& operator<<(std::ostream& out, const hardware_platform value){
+    static std::map<hardware_platform, std::string> strings;
+    if (strings.size() == 0){
+        strings[hardware_platform::cpu] = "cpu";
+        strings[hardware_platform::cuda] = "cuda";
+        strings[hardware_platform::rocm] = "rocm";
```

Comment:
please decide on one approach. Either with `INSERT_ELEMENT()` or without. I think we said that `INSERT_ELEMENT` is not worth it, if I remember correctly..